Add coupler support

nanosonde opened the pull request:
This PR adds coupler support. More details here: https://github.com/thelsing/knx/issues/76

thelsing commented on examples/knx-linux-coupler/main.cpp:
```diff
@@ -0,0 +1,112 @@
+    calc.snprintFdsk(fdskString, sizeof(fdskString), serialNumber, key);
+    printf("FDSK: %s\n", fdskString);
+
+    // Prevent swapping of this process
```
Just out of interest: why is it necessary to prevent swapping?

nanosonde replied:
This was introduced when I added KNX-RF. As we only have the single main loop and RF timing is critical, I have added this to make sure that any process memory is swapped to disk.

nanosonde replied:
I have no proof that this really has any effect. As KNX-RF with a CC1101 directly connected via SPI is really ugly. The RPi causes so much interference the RX sensitivity is really bad. Ok, with a proper PCB layout and some filtering this could be handled, but I am not that much interested in using the CC1101 directly with the RPi. It served the purpose of being a dev platform to test things quickly related to the CC1101.

thelsing commented on src/knx/tpuart_data_link_layer.h:
```diff
@@ -8,19 +8,26 @@
 
 #define MAX_KNX_TELEGRAM_SIZE 263
 
+class ITpUartCallBacks
```
I would have use a function pointer for this. But I don't have a strong opinion about it. Not sure whats better. A function pointer may be a bit faster. Any opinion on this? I would like to have callbacks handled in a uniform way across the project.

nanosonde replied:
Normally I would have used std::function for this. I have already found some implementation which do not use dynamic memory for this.
Example: https://github.com/arobenko/embxx/blob/master/embxx/util/StaticFunction.h

nanosonde replied:
This was my try: https://github.com/nanosonde/knx/blob/feat_coupler/src/knx/simple_functional.h
But I have decided not to use it as it seems to be a mess.

thelsing replied:
So simple_functional is not used? Then it shouldn't be in the pull request (and I don't have to try to understand it ;)

nanosonde replied:
Yes, sorry. I forgot to remove it.

nanosonde replied:
Maybe we should checkout this implementation: https://github.com/arobenko/embxx/blob/master/embxx/util/StaticFunction.h
It would have to be adapted though to replace the GASSERT().

nanosonde replied:
It will still require the header #include <functional>

nanosonde replied:
Please also read this: https://arobenko.gitbooks.io/bare_metal_cpp/content/basic_needs/callback.html

thelsing replied:
I'd rather keep it as it is. I don't really see why we should use a complicated template. The current solution is much easier to understand.

nanosonde replied:
Ok. However, after thinking about it, I would say that the implementation should be changed.
It is the responsibilty of the network layer to give an answer if a packet should be ACK'ed or not.
For a router and also for normal device. So instead of placing it in the BAU, I think there should be an interface via the network entity to ask the network layer "shallPackedBeAcked". Then there is no need for registering a callback. It is then part of the normal network layer (entity) interface towards the datalink layer.
What do you think?

nanosonde replied:
Note: If a datalink layer does not ACK packets then it simply does not ask the network layer. E.g.: IP and RF (more specific RF ready, RF multi can ACK packets)

thelsing replied:
Feel free to change it.

nanosonde replied:
I will change this. Stay tuned.

nanosonde replied:
After thinking about it again, I will leave it as is for now as it only affects TP data link layer which uses ACKs.

nanosonde commented:
Just a small note:
I have decided to NOT split up the TransportLayer and ApplicationLayer classes each of it into two parts: one for the management stuff and one for the group comm stuff as I did it for the bau_SystemB class.
Therefore I made group comm related things optional by using a pointer which could be nullptr.

thelsing commented:
You're to fast. I still have to look over 50 files or so.

nanosonde commented:
Yeah, ok. I will wait with my comments. :-)

thelsing commented on src/knx/bau57B0.cpp:
```diff
@@ -9,16 +9,17 @@
 #endif
 {
-    _netLayer.dataLinkLayer(_dlLayer);
+    _netLayer.getEntity(0).dataLinkLayer(_dlLayer);
```
Just to nitpick: One could eliminate the magic number 0 ;)

thelsing replied:
Same with all other getEntity calls. Are there more than 2 of them? If not getPrimary and getSecondary are easier to understand.

nanosonde replied:
The current KNX spec. does not forsee more than two interfaces, although they mention in the coupler model 2.0 spec that it could be an option in the future.
I will add public members getPrimary() and getSecondary().

nanosonde replied:
replaced.

thelsing reviewed (approved):
Seems fine to me. Some comments:
- documentation: can you a bit documentation how to use it? Is it possible to use ESP8266 with TPUART as an IP/TP Router? It would be nice if you could update the class diagramm (Same for the secure stuff)
- you introduced the MASK-defines. We should use the in the baus to only compile the bau if its MASK is defined
- for further pull requests: It would've been helpful if you described you changes in a few sentences. 

Nothing to keep this from getting merged. Just nice to have. What do you say?

nanosonde commented:
Could you please check again?
I have added some minimal docs, updated the class diagram for normal devices and added class diagram for a IP/TP1 coupler.
Also changed the things you had requested except the callback for TPUART ACKs.

Travis CI also modified. ESP8266 commented out as I had problems with it: https://github.com/thelsing/knx/issues/86

thelsing commented:
Merged. Thanks as always.
